src/generic_term.cpp

```diff
@@ -0,0 +1,196 @@
+/*********************                                                        */
+/*! \file generic_term.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Yoni Zohar
+** This file is part of the smt-switch project.
+** Copyright (c) 2020 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief Class that maintains the expected
+**        Op and children
+**
+**/
+
+#include "generic_term.h"
+
+#include "exceptions.h"
+#include "utils.h"
+
+using namespace std;
+
+namespace smt {
+
+/* GenericTerm */
+
+GenericTerm::GenericTerm(Sort s, Op o, TermVec c, string r)
+    : sort(s), op(o), children(c), repr(r), is_sym(false), is_par(false)
+{
+  ground = compute_ground();
+}
+
+GenericTerm::GenericTerm(Sort s, Op o, TermVec c, string r, bool is_sym)
+    : sort(s),
+      op(o),
+      children(c),
+      repr(r),
+      is_sym(is_sym),
+      is_par(!is_sym),
+      ground(true)
+{
+}
+
+bool GenericTerm::compute_ground()
+{
+  // variables are not ground terms
+  if (is_param())
+  {
+    return false;
+  }
+  // a compound term is ground iff all
+  // of its children are ground
+  for (Term child : get_children())
+  {
+    shared_ptr<GenericTerm> gc = static_pointer_cast<GenericTerm>(child);
+    // Note that this is not recursive -- is_ground
+    // is just a getter.
+    if (!gc->is_ground())
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+GenericTerm::~GenericTerm() {}
+
+Op GenericTerm::get_op() const { return op; }
+
+TermVec GenericTerm::get_children() { return children; }
+
+Sort GenericTerm::get_sort() const { return sort; }
+
+std::size_t GenericTerm::get_id() const { return hash(); }
+
+bool GenericTerm::compare(const Term & t) const
+{
+  // TODO: not efficient. Should do similar to logging term.
+  // the difference is that there is no underlying solver,
+  // and that we need some string sometimes anyway to create a
+  // generic term.
+  if (!t)
+  {
+    // The null term is different than any constructed term.
+    return false;
+  }
+  shared_ptr<GenericTerm> gt = static_pointer_cast<GenericTerm>(t);
+  // The comparison is based on a string comparison
+  return repr == gt->to_string();
+}
+
+string GenericTerm::compute_string() const
+{
+  if (!repr.empty())
+  {
+    return repr;
+  }
+  Assert(!op.is_null());
```

Comment:
Just checking, does this work because symbols always have their `repr` set on construction?

Reply:
yes, that's the idea. although i agree it is not super clear from the code right now. I am open for suggestion if there is a cleaner way for this.